model/riscv_vmem_utils.sail

```diff
@@ -0,0 +1,155 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+/* This file implements utility functions for accessing memory that
+ * can be used by instruction definitions.
+ */
+
+/* This is a external option that controls the order in which the model
+ * performs misaligned accesses.
+ */
+val sys_misaligned_order_decreasing = {
+  ocaml: "Platform.misaligned_order_decreasing",
+  _: "sys_misaligned_order_decreasing"
+} : unit -> bool
+
+/* This is an external option that, when true, causes all misaligned accesses
+ * to be split into single byte operations.
+ */
+val sys_misaligned_to_byte = {
+  ocaml: "Platform.misaligned_to_byte",
+  _: "sys_misaligned_to_byte"
+} : unit -> bool
+
+val split_misaligned : forall 'width, 'width in {1, 2, 4, 8}.
```

Comment:
Unfortunately there's also `cbo.zero` which does 32 byte writes (by default). I think you can do this without a `match` anyway though.

```
val count_trailing_zeros : forall 'n, 'n >= 0. (bits('n)) -> range(0, 'n)
function count_trailing_zeros(x) = {
    foreach (i from 0 to ('n - 1)) {
        if x[i] == bitone then return i;
    };
    'n
}

val split_misaligned : forall 'width, 'width in {1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096} .
  (xlenbits, int('width)) -> {'n 'bytes, 'width == 'n * 'bytes & 'bytes > 0. (int('n), int('bytes))}
function split_misaligned(vaddr, width) = {
    if is_aligned_addr(vaddr, width) then (1, width)
    else if sys_misaligned_to_byte() then (width, 1)
    else {
        let vaddr_alignment_bytes = 2 ^ count_trailing_zeros(vaddr);
        assert(vaddr_alignment_bytes <= width);
        (vaddr_alignment_bytes, width / vaddr_alignment_bytes)
    }
}
```

I couldn't figure out how to get "must be a power of two" into the type system though. :-(

Reply:
You can't, as it would require an existential quantifier (M is a power of two becomes 'there exists an N such that M = 2 ^ N'). We could add some syntactic sugar for a set of powers of two up to some bound though.